src/main/java/com/jeovan/gymcrmsystem/services/TraineeServiceImpl.java

```diff
@@ -2,49 +2,91 @@
+import com.jeovan.gymcrmsystem.models.Trainer;
+import com.jeovan.gymcrmsystem.models.User;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.annotation.Secured;
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.config.annotation.authentication.configuration.EnableGlobalAuthentication;
+import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.stereotype.Service;
 
 import java.util.List;
-import java.util.Map;
 import java.util.UUID;
 
 @Service
 public class TraineeServiceImpl implements TraineeService {
-    @Autowired
-    private InMemoryStorage inMemoryStorage;
-
     @Autowired
     private TraineeDao traineeDao;
+    @Autowired
+    private CredentialGeneratorServiceImpl credentialGeneratorService;
+    @Autowired
+    private PasswordEncoder passwordEncoder;
 
     @Override
-    public void setTraineeService() {
-        traineeDao.setStorage((Map<UUID, Trainee>) inMemoryStorage.getStorage().get("Trainee"));
-    }
-
-    @Override
-    public Map<UUID, Trainee> getAll() {
-        return traineeDao.getAll(Trainee.class.getSimpleName());
+    @Secured("ADMIN")
+    public List<Trainee> getAll() {
+        return traineeDao.findAll();
     }
 
     @Override
     public Trainee create(Trainee trainee) {
+        User user = trainee.getUser();
+        user.setUsername(credentialGeneratorService.generateUsername(user.getFirstName(), user.getLastName()));
+        user.setPassword(passwordEncoder.encode(credentialGeneratorService.generatePassword()));
         return traineeDao.save(trainee);
     }
 
     @Override
+    @Secured("ADMIN")
     public Trainee update(Trainee trainee) {
-        return traineeDao.update(trainee);
+        return traineeDao.save(trainee);
     }
 
     @Override
+    @Secured("ADMIN")
     public Trainee select(UUID id) {
-        return traineeDao.getById(id).get();
+        return traineeDao.findById(id).get();
+    }
+
+    @Secured("ADMIN")
+    public Trainee selectByUsername(String username) {
+        return traineeDao.findByUsername(username).get();
     }
 
+    @Secured("ADMIN")
+    public Trainee updatePassword(String username){
+        Trainee trainee = selectByUsername(username);
+        User user = trainee.getUser();
+        user.setPassword(passwordEncoder.encode(credentialGeneratorService.generatePassword()));
+        return traineeDao.save(trainee);
+    }
+
+    @Secured("ADMIN")
+    public Trainee toggleActiveStatus(String username){
```

Comment:
I think toggleActiveStatus should be part of the interface so we can use it in the future regardless of the implementation

Reply:
Added to interface